text/0000-mixed-shadow-mode.md

```diff
@@ -0,0 +1,199 @@
+of these polyfills can result in broken experiences due to dependencies on certain "escape hatches"
+that are not possible in a native Shadow DOM environment (e.g., instrumentation, styling, etc, which
+are discussed below).
+
+Applications that wish to use LWC components in a native shadow context simply need to omit the
+inclusion of the `@lwc/synthetic-shadow` polyfill; however, this is not a realistic option for
+large applications that cannot afford a rewrite.
+
+This proposal opens up an incremental migration path for applications moving to native Shadow DOM by
+allowing the usage of both native and synthetic Shadow DOM in the same application.
+
+## Detailed design
+
+Up until now, the two choices of shadow semantics have been synthetic mode and native mode. All the
+components in an application run in either synthetic mode or native mode, depending on whether the
+synthetic shadow polyfill (i.e., `@lwc/synthetic-shadow`) is applied. Mixed mode introduces the
+ability for a component to access native Shadow DOM APIs, even if the synthetic shadow polyfill is
+applied. Mixed mode should be transparent to components and the framework should guarantee this
+through integration testing, which is further discussed below.
+
+A component can signal that it prefers native Shadow DOM by setting a static property called
+`defaultShadowMode` to `native`.
+
+```js
+export default class extends LightningElement {
+    static defaultShadowMode = 'native';
+}
+```
+
+If this property is not defined, a default value of `false` will be used. The value of this property
+will be read before component construction and cached for the lifetime of the application. The LWC
+engine will throw an exception if the value of `defaultShadowMode` is unrecognized.
+
+All polyfills applied by `@lwc/synthetic-shadow` will look to the value of this property to choose
+whether to invoke the polyfilled API or the native browser API. If the browser does not support
+Shadow DOM (e.g., IE11), then LWC will fallback to synthetic mode regardless of which mode the
+component prefers.
+
+Due to observable differences between the two modes, it is likely that components that prefer native
+will need to know whether it is currently operating in native or synthetic Shadow DOM. Any hints
+provided by the framework will only be available to components that set the `defaultShadowMode`
+property to `native` because this information is not useful to components that only support
+synthetic Shadow DOM.
+
+In terms of composition, a synthetic mode component can contain a native mode component, but the
+inverse is not allowed. Not only does this make things easier to reason about, but many existing
+components rely on workarounds in synthetic mode that are not possible to support in native mode.
+These observable differences are further discussed below. As this invariant cannot be asserted
+during compile-time, the engine will assert it during runtime.

```

Comment:
We're starting to discuss the complexities of implementing the synthetic-within-native restriction (e.g. how to handle synthetic-within-light-within-native, how to handle dynamic DOM node insertion), so I think it would be good to spell out exactly what benefits we get from restricting synthetic-within-native.

Also: why restrict synthetic-within-native, but not native-within-synthetic?

Reply:
this is not an arbitrary restriction, this is a technical driven restriction, any slot element printed by a synthetic shadow will be inherited (owned) by the nearest document fragment, which is going to be native, and mess up with the distribution for that native.

Reply:
Thanks, this makes it much clearer why we should restrict it! 🙂  I can't think of a workaround, unless we were to transform `<slot>` into something else at compile time when it's a synthetic shadow component.

Reply:
which will be a non-backward compatible change for that context, breaking some components, so better to just restrict it.